src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java

```diff
@@ -0,0 +1,161 @@
+
+	@Autowired
+	private BreedingMethodService breedingMethodService;
+
+	public void validateEmptyList(final BindingResult errors, final List<GermplasmUpdateDTO> germplasmUpdateDTOList) {
+		if (germplasmUpdateDTOList == null || germplasmUpdateDTOList.isEmpty()) {
+			errors.reject("germplasm.update.empty.list", "");
+		}
+	}
+
+	public void validateCodes(final BindingResult errors, final List<GermplasmUpdateDTO> germplasmUpdateDTOList) {
+
+		final Set<String> nameCodes = new HashSet<>();
+		germplasmUpdateDTOList
+			.forEach(g -> nameCodes.addAll(g.getNames().keySet().stream().map(n -> n.toUpperCase()).collect(Collectors.toList())));
+		nameCodes.addAll(
+			germplasmUpdateDTOList.stream()
+				.map(o -> StringUtils.isNotEmpty(o.getPreferredName()) ? o.getPreferredName().toUpperCase() : null)
+				.filter(Objects::nonNull).collect(
+				Collectors.toSet()));
+		final List<String> existingNamesCode =
+			this.germplasmService.filterGermplasmNameTypes(nameCodes).stream().map(GermplasmNameTypeDTO::getCode).collect(
+				Collectors.toList());
+
+		final Set<String> attributesCode = new HashSet<>();
+		germplasmUpdateDTOList.stream().filter(germ -> germ.getAttributes() != null).collect(Collectors.toList())
+			.forEach(
+				g -> attributesCode.addAll(g.getAttributes().keySet().stream().map(n -> n.toUpperCase()).collect(Collectors.toList())));
+		final List<String> existingAttributesCode =
+			this.germplasmService.filterGermplasmAttributes(attributesCode).stream().map(AttributeDTO::getCode).collect(
+				Collectors.toList());
+
+		nameCodes.removeAll(existingNamesCode);
+		attributesCode.removeAll(existingAttributesCode);
+
+		if (!nameCodes.isEmpty()) {
+			errors.reject("germplasm.update.invalid.name.code", new String[] {String.join(",", nameCodes)}, "");
+		}
+		if (!attributesCode.isEmpty()) {
+			errors.reject("germplasm.update.invalid.attribute.code", new String[] {String.join(",", attributesCode)}, "");
+		}
+
+	}
+
+	public void validateGermplasmIdAndGermplasmUUID(final BindingResult errors, final List<GermplasmUpdateDTO> germplasmUpdateDTOList) {
+		// Find rows (GermplasmUpdateDTO) with blank GID and UUID.
+		final Optional<GermplasmUpdateDTO> optionalGermplasmUpdateDTO =
+			germplasmUpdateDTOList.stream().filter(dto -> StringUtils.isEmpty(dto.getGermplasmUUID()) && dto.getGid() == null).findAny();
+		if (optionalGermplasmUpdateDTO.isPresent()) {
+			errors.reject("germplasm.update.missing.gid.and.uuid", "");
+		}
+
+		final Set<Integer> gids =
+			germplasmUpdateDTOList.stream().map(dto -> dto.getGid()).filter(Objects::nonNull).collect(Collectors.toSet());
+		final Set<String> germplasmUUIDs =
+			germplasmUpdateDTOList.stream().map(dto -> StringUtils.isNotEmpty(dto.getGermplasmUUID()) ? dto.getGermplasmUUID() : null)
+				.filter(
+					Objects::nonNull).collect(Collectors.toSet());
+
+		final List<Germplasm> germplasmByGIDs = this.germplasmDataManager.getGermplasms(new ArrayList<>(gids));
```

Comment:
I suggest we implement` GermplasmService#getGermplasmByGIDs` and so as to reduce usage and dependency on `GermplasmDataManager` in BMSAPI